wordpress-mobile/WordPress-iOS
# [Tooling] Add linter for AppLocalizedString not used in extension targets

This work is part of paaHJt-2Ib-p2, and a follow up of the annoucement in paNNhX-nP-p2.

## What

Adds a linter which checks that every source file included in an app extension target (e.g. Widgets, NotificationExtension, ShareExtension…) uses `AppLocalizedString` instead of `NSLocalizedString`, to be sure it uses the translations from the app bundle (as we put all our `Localizable.strings` file in the app bundle to share between them app and extension targets, and reduce the overall `.ipa` size)

This also fixes 3 of such violations that the linter found in the process.

## Screenshot of the linter in action

<img width="1493" alt="image" src="https://user-images.githubusercontent.com/216089/164788414-ed717f39-071b-4d81-880b-7b88f59c4328.png">

<details><summary>Another example on a file used in 4 different app extension targets</summary>
<img width="1494" alt="image" src="https://user-images.githubusercontent.com/216089/164782794-b0581660-56ae-49ff-b7d8-81f06bcd1d5e.png">
</details>

## Related Past PRs

For reference, the following past PRs merged a while ago migrated call sites of `NSLocalizedString` to `AppLocalizedString` for files used in extension targets.

- https://github.com/wordpress-mobile/WordPress-iOS/pull/17570
- https://github.com/wordpress-mobile/WordPress-iOS/pull/17630
- https://github.com/wordpress-mobile/WordPress-iOS/pull/17636

This is why the linter that this PR adds today only found 3 violations (introduced after these PRs landed) and not hundreds.

## To Test

 - 🔨 Build the `WordPress` target in Xcode, and verify that there is no error (CI will already do that for you)
 - ✏️  Edit the `WordPress/Classes/ViewRelated/Stats/Extensions/Double+Stats.swift` file and replace one of the `AppLocalizedString` calls it uses with `NSLocalizedString`
 - 🔨 Build the `WordPressShareExtension` scheme, and verify that the build fails and the error is shown in on the right line.
   - 🕵️ There might be multiple errors on the same line (one for each extension target complaining about it) if that file is build for multiple extension targets
   - 🕵️ For example, when building the `WordPressShareExtension` scheme, Xcode actually also builds the `WordPress` app target (given how that scheme is configured see Edit Scheme > Build tab), which in turns builds other app extensions that the `.app` will embed in its bundle. Hence potentially many more targets being built depending on the schemes, and thus more instances of the error
 - 💻  Optional: run `Scripts/BuildPhases/LintAppLocalizedStringsUsage.rb` in your Terminal and observe that it will scan all the App Extension targets found in the project — and report the same errors as the ones in Xcode — and exits with status code 1 (`echo $?`)
 - ✏️  Fix the violation you introduced (i.e. revert the call you changed)
 - 🔨 Start a build again, and verify it still passes.
